libvast/src/table_slice.cpp

```diff
@@ -891,6 +883,263 @@ auto resolve_operand(const table_slice& slice, const operand& op)
+    },
+    [&](const record_type& rt) -> indexed_transformation::result_type {
+      return flatten_record(opt, name_prefix, list_offsets, {field.name, rt},
+                            list_array->values());
+    },
+  };
+  return caf::visit(f, lt.value_type());
+}
+
+auto flatten_record(
+  const flatten_options& opt, std::string_view name_prefix,
+  const std::vector<std::shared_ptr<arrow::Array>>& list_offsets,
+  struct record_type::field field, const std::shared_ptr<arrow::Array>& array)
+  -> indexed_transformation::result_type {
+  const auto& rt = caf::get<record_type>(field.type);
+  auto struct_array
+    = std::static_pointer_cast<type_to_arrow_array_t<record_type>>(array);
+  const auto next_name_prefix
+    = fmt::format("{}{}{}", name_prefix, field.name, opt.separator);
+  auto transformations = std::vector<indexed_transformation>{};
+  for (size_t i = 0; i < rt.num_fields(); ++i) {
+    transformations.push_back(
+      {offset{i},
+       make_flatten_transformation(opt, next_name_prefix, list_offsets)});
+  }
+  auto [output_type, output_struct_array]
+    = transform_columns(field.type, struct_array, transformations);
+  VAST_ASSERT(output_type);
+  VAST_ASSERT(output_struct_array);
+  auto result = indexed_transformation::result_type{};
+  result.reserve(output_struct_array->num_fields());
+  const auto& output_rt = caf::get<record_type>(output_type);
+  for (int i = 0; i < output_struct_array->num_fields(); ++i) {
+    const auto field_view = output_rt.field(i);
+    result.push_back({
+      {std::string{field_view.name}, field_view.type},
+      output_struct_array->field(i),
+    });
+  }
+  return result;
+}
+
+auto make_flatten_transformation(
+  const flatten_options& opt, const std::string& name_prefix,
+  std::vector<std::shared_ptr<arrow::Array>> list_offsets)
+  -> indexed_transformation::function_type {
+  return [=](struct record_type::field field,
+             std::shared_ptr<arrow::Array> array)
+           -> indexed_transformation::result_type {
+    auto f = detail::overload{
+      [&]<concrete_type Type>(
+        const Type&) -> indexed_transformation::result_type {
+        // Return unchanged, but use prefix, and wrap in a list if we need to.
+        if (list_offsets.empty()) {
+          return {
+            {
+              {
+                fmt::format("{}{}", name_prefix, field.name),
+                field.type,
+              },
+              array,
+            },
+          };
+        }
+        auto combined_offsets = combine_offsets(list_offsets);
+        return {
+          {
+            {
+              fmt::format("{}{}", name_prefix, field.name),
+              type{list_type{field.type}},
+            },
+            arrow::ListArray::FromArrays(*combined_offsets, *array).ValueOrDie(),
+          },
+        };
+      },
+      [&](const list_type&) -> indexed_transformation::result_type {
+        if (opt.ignore_lists) {
```

Comment:
I think this does not work for records inside lists inside lists. Consider this:

```json5
// Input
{"foo": [[{"bar": "baz"}]]}

// Without -i
{"foo": "bar": ["baz"]}

// Actual output with -i
{"foo": [[{"bar": "baz"}]]}

// Expected output with -i
{"foo": "bar": [["baz"]]}
```

Reply:
It "works" insofar that `ignore-lists` ignores lists including their values.
Should `ignore-lists` not ignore record values in lists, but any other scalar and list value?

Reply:
Yes. We could also move this into a follow-up story, but then we should remove the option for now (maybe make that a separate commit so that can be reverted as a starting point).

Reply:
Alright, I tried moving the `if (opt.ignore_lists)` around in the recursive call chain but that didn't achieve anything.

I removed `ignore-lists` in https://github.com/tenzir/tenzir/pull/3108/commits/a1298a1668f20052a1500a22289b0be8c3580199.